File: Source/ARTCrypto.m
//
//  ARTCrypto.m
//  ably-ios
//
//  Created by Jason Choy on 20/12/2014.
//  Copyright (c) 2014 Ably. All rights reserved.
//

#import "ARTCrypto+Private.h"

#import <CommonCrypto/CommonCrypto.h>

#define ART_CBC_BLOCK_LENGTH (16)

@interface ARTCipherParams ()

- (BOOL)ccAlgorithm:(CCAlgorithm *)algorithm error:(NSError **)error;

@end

@interface ARTCrypto ()

@property (nonatomic, weak) ARTLog * logger;

@end

@interface ARTCbcCipher ()

@property CCAlgorithm algorithm;

@end

@implementation NSString (ARTCipherKeyCompatible)

- (NSData *)toData {
    NSString *key = self;
    key = [key stringByReplacingOccurrencesOfString:@"-" withString:@"+"];
    key = [key stringByReplacingOccurrencesOfString:@"_" withString:@"/"];
    return [[NSData alloc] initWithBase64EncodedString:key options:0];
}

@end

@implementation NSData (ARTCipherKeyCompatible)

- (NSData *)toData {
    return self;
}

@end

@implementation ARTCipherParams

- (instancetype)initWithAlgorithm:(NSString *)algorithm key:(id<ARTCipherKeyCompatible>)key {
    NSData *keyData = [key toData];
    return [self initWithAlgorithm:algorithm key:keyData iv:nil];
}

- (instancetype)initWithAlgorithm:(NSString *)algorithm key:(id<ARTCipherKeyCompatible>)key iv:(NSData *)iv {
    self = [super init];
    if (self) {
        _algorithm = algorithm;
        _key = [key toData];
        _keyLength = [_key length] * 8;
        _iv = iv;

        CCAlgorithm ccAlgorithm;
        NSError *error = nil;
        if (![self ccAlgorithm:&ccAlgorithm error:&error]) {
            [ARTException raise:NSInvalidArgumentException format:@"%@", error.userInfo[NSLocalizedFailureReasonErrorKey]];
        }
    }
    return self;
}

- (NSString *)getMode {
    return @"CBC";
}

- (BOOL)ccAlgorithm:(CCAlgorithm *)algorithm error:(NSError **)error {
    NSString *errorMsg;
    if (NSOrderedSame == [self.algorithm compare:@"AES" options:NSCaseInsensitiveSearch]) {
        if (self.iv != nil && [self.iv length] != ART_CBC_BLOCK_LENGTH) {
            errorMsg = [NSString stringWithFormat:@"iv length expected to be %d, got %d instead", ART_CBC_BLOCK_LENGTH, (int)[self.iv length]];
        } else if (self.keyLength != 128 && self.keyLength != 256) {
            errorMsg = [NSString stringWithFormat:@"invalid key length for AES algorithm: %d", (int)self.keyLength];
        } else {
            *algorithm = kCCAlgorithmAES128;
        }
    } else if (NSOrderedSame == [self.algorithm compare:@"DES" options:NSCaseInsensitiveSearch]) {
        *algorithm = kCCAlgorithmDES;
    } else if (NSOrderedSame == [self.algorithm compare:@"3DES" options:NSCaseInsensitiveSearch]) {
        *algorithm = kCCAlgorithm3DES;
    } else if (NSOrderedSame == [self.algorithm compare:@"CAST" options:NSCaseInsensitiveSearch]) {
        *algorithm = kCCAlgorithmCAST;
    } else if (NSOrderedSame == [self.algorithm compare:@"RC4" options:NSCaseInsensitiveSearch]) {
        *algorithm = kCCAlgorithmRC4;
    } else if (NSOrderedSame == [self.algorithm compare:@"RC2" options:NSCaseInsensitiveSearch]) {
        *algorithm = kCCAlgorithmRC2;
    } else {
        errorMsg = [NSString stringWithFormat:@"unknown algorithm: %@", self.algorithm];
    }

    if (errorMsg) {
        [self.logger error:@"ARTCrypto.ccAlgorithm: %@", errorMsg];
        if (error) *error = [NSError errorWithDomain:ARTAblyErrorDomain code:0 userInfo:@{NSLocalizedFailureReasonErrorKey: errorMsg}];
        return NO;
    }

    return YES;
}

- (ARTCipherParams *)toCipherParams {
    return self;
}

@end

@implementation NSDictionary (ARTCipherParamsCompatible)

- (ARTCipherParams *)toCipherParams {
    return [ARTCrypto getDefaultParams:self];
}

@end

@implementation ARTCbcCipher

- (id)initWithCipherParams:(ARTCipherParams *)cipherParams {
    self = [super init];
    if (self) {
        _keySpec = cipherParams.key;
        _iv = cipherParams.iv;
        _blockLength = ART_CBC_BLOCK_LENGTH;

        if (![cipherParams ccAlgorithm:&_algorithm error:nil]) {
            return nil;
        }
    }
    return self;
}

-(size_t) keyLength {
    return [self.keySpec length] *8;
}

+ (instancetype)cbcCipherWithParams:(ARTCipherParams *)cipherParams {
    return [[self alloc] initWithCipherParams:cipherParams];
}

- (ARTStatus *)encrypt:(NSData *)plaintext output:(NSData *__autoreleasing *)output {
    NSData *iv = self.iv != nil ? self.iv : [ARTCrypto generateSecureRandomData:self.blockLength];
    NSData *ciphertext = nil;

    // The maximum cipher text is plaintext length + block length. We are also prepending this with the IV so need 2 block lengths in addition to the plaintext length.
    size_t outputBufLen = [plaintext length] + self.blockLength * 2;
    void *buf = malloc(outputBufLen);

    if (!buf) {
        [self.logger error:@"ARTCrypto error encrypting"];
        return [ARTStatus state:ARTStateError];
    }

    // Copy the iv first
    memcpy(buf, [iv bytes], self.blockLength);

    void *ciphertextBuf = ((char *)buf) + self.blockLength;
    size_t ciphertextBufLen = outputBufLen - self.blockLength;

    const void *key = [self.keySpec bytes];
    size_t keyLen = [self.keySpec length];

    const void *ivBytes = [iv bytes];
    const void *dataIn = [plaintext bytes];
    size_t dataInLen = [plaintext length];

    size_t bytesWritten = 0;
    CCCryptorStatus status = CCCrypt(kCCEncrypt, self.algorithm, kCCOptionPKCS7Padding, key, keyLen, ivBytes, dataIn, dataInLen, ciphertextBuf, ciphertextBufLen, &bytesWritten);

    if (status) {
        [self.logger error:@"ARTCrypto error encrypting. Status is %d", status];
        free(ciphertextBuf);
        return [ARTStatus state: ARTStateError];
    }

    ciphertext = [NSData dataWithBytesNoCopy:buf length:(bytesWritten + self.blockLength) freeWhenDone:YES];
    if (nil == ciphertext) {
        [self.logger error:@"ARTCrypto error encrypting. cipher text is nil"];
        free(buf);
        return [ARTStatus state:ARTStateError];
    }

    *output = ciphertext;

    return [ARTStatus state:ARTStateOk];
}

- (ARTStatus *)decrypt:(NSData *)ciphertext output:(NSData *__autoreleasing *)output {
    // The first *blockLength* bytes are the iv
    if ([ciphertext length] < self.blockLength) {
        return [ARTStatus state: ARTStateInvalidArgs];;
    }

    NSData *ivData = [ciphertext subdataWithRange:NSMakeRange(0, self.blockLength)];
    NSData *actualCiphertext = [ciphertext subdataWithRange:NSMakeRange(self.blockLength, [ciphertext length] - self.blockLength)];

    CCOptions options = 0;
    const void *key = [self.keySpec bytes];
    size_t keyLength = [self.keySpec length];

    const void *iv = [ivData bytes];
    const void *dataIn = [actualCiphertext bytes];
    size_t dataInLength = [actualCiphertext length];

    // The output will never be more than the input + block length
    size_t outputLength = dataInLength + self.blockLength;
    void *buf = malloc(outputLength);
    size_t bytesWritten = 0;

    if (!buf) {
        [self.logger error:@"ARTCrypto error decrypting."];
        return [ARTStatus state:ARTStateError];
    }

    // Decrypt without padding because CCCrypt does not return an error code
    // if the decrypted value is not padded correctly
    CCCryptorStatus status = CCCrypt(kCCDecrypt, self.algorithm, options, key, keyLength, iv, dataIn, dataInLength, buf, outputLength, &bytesWritten);

    if (status) {
        [self.logger error:@"ARTCrypto error decrypting. Status is %d", status];
        free(buf);
        return [ARTStatus state:ARTStateError];
    }

    // Check that the decrypted value is padded correctly and determine the unpadded length
    const char *cbuf = (char *)buf;
    int paddingLength = cbuf[bytesWritten - 1];

    if (0 == paddingLength || paddingLength > bytesWritten) {            free(buf);
        return [ARTStatus state:ARTStateCryptoBadPadding];
    }

    for (size_t i=(bytesWritten - 1); i>(bytesWritten - paddingLength); --i) {
        if (paddingLength != cbuf[i-1]) {
            free(buf);
            return [ARTStatus state:ARTStateCryptoBadPadding];
        }
    }

    size_t unpaddedLength = bytesWritten - paddingLength;

    NSData *plaintext = [NSData dataWithBytesNoCopy:buf length:unpaddedLength freeWhenDone:YES];
    if (!plaintext) {
        [self.logger error:@"ARTCrypto error decrypting. plain text is nil"];
        free(buf);
    }

    *output = plaintext;

    return [ARTStatus state:ARTStateOk];
}

- (NSString *)cipherName {
    NSString *algo = nil;
    switch (self.algorithm) {
        case kCCAlgorithmAES128:
            algo = @"aes-128";
            break;
        case kCCAlgorithmDES:
            algo = @"des";
            break;
        case kCCAlgorithm3DES:
            algo = @"3des";
            break;
        case kCCAlgorithmCAST:
            algo = @"cast";
            break;
        case kCCAlgorithmRC4:
            algo = @"rc4";
            break;
        case kCCAlgorithmRC2:
            algo = @"rc2";
            break;
        default:
            NSAssert(NO, @"Invalid algorithm");
            return nil;
    }
    return [NSString stringWithFormat:@"%@-cbc", algo];
}

@end

@implementation ARTCrypto

+ (NSString *)defaultAlgorithm {
    return @"AES";
}

+ (int)defaultKeyLength {
    return 256;
}

+ (int)defaultBlockLength {
    return 128;
}

+ (NSMutableData *)generateSecureRandomData:(size_t)length {
    void *buf = malloc(length);
    if (!buf) {
        return nil;
    }
    int rc = SecRandomCopyBytes(kSecRandomDefault, length, buf);
    if (rc != 0) {
        free(buf);
        return nil;
    }

    NSMutableData *outputData = [NSMutableData dataWithBytesNoCopy:buf length:length freeWhenDone:YES];
    if (!outputData) {
        free(buf);
    }
    return outputData;
}

+ (ARTCipherParams *)getDefaultParams:(NSDictionary *)values {
    NSString *algorithm = values[@"algorithm"];
    if (algorithm == nil) {
        algorithm = [ARTCrypto defaultAlgorithm];
    }
    NSString *key = values[@"key"];
    if (key == nil) {
        [ARTException raise:NSInvalidArgumentException format:@"missing key parameter"];
    }
    return [[ARTCipherParams alloc] initWithAlgorithm:algorithm key:key];
}

+ (NSData *)generateRandomKey {
    return [self generateRandomKey:[self defaultKeyLength]];
}

+ (NSData *)generateRandomKey:(NSUInteger)length {
    return [self generateSecureRandomData:length / 8];
}

+ (id<ARTChannelCipher>)cipherWithParams:(ARTCipherParams *)params {
    return [ARTCbcCipher cbcCipherWithParams:params];
}

@end

File: Sources/AblySwift/ARTCrypto.swift
import Foundation
import CommonCrypto
import Security

// swift-migration: original location ARTCrypto.m, line 6
private let cbcBlockLength = 16

// swift-migration: original location ARTCrypto.h, line 9 and ARTCrypto.m, line 9
public protocol ARTCipherKeyCompatible {
    func toData() -> Data
}

// swift-migration: original location ARTCrypto.h, line 14 and ARTCrypto.m, line 20
extension String: ARTCipherKeyCompatible {
    
    // swift-migration: original location ARTCrypto.h, line 15 and ARTCrypto.m, line 22
    public func toData() -> Data {
        var key = self
        key = key.replacingOccurrences(of: "-", with: "+")
        key = key.replacingOccurrences(of: "_", with: "/")
        return Data(base64Encoded: key) ?? Data()
    }
}

// swift-migration: original location ARTCrypto.h, line 19 and ARTCrypto.m, line 31
extension Data: ARTCipherKeyCompatible {
    
    // swift-migration: original location ARTCrypto.h, line 20 and ARTCrypto.m, line 33
    public func toData() -> Data {
        return self
    }
}

// swift-migration: original location ARTCrypto.h, line 26
public protocol ARTCipherParamsCompatible {
    func toCipherParams() -> ARTCipherParams
}

// swift-migration: original location ARTCrypto.h, line 31 and ARTCrypto.m, line 105
extension Dictionary: ARTCipherParamsCompatible where Key == String, Value == Any {
    
    // swift-migration: original location ARTCrypto.h, line 32 and ARTCrypto.m, line 107
    public func toCipherParams() -> ARTCipherParams {
        return ARTCrypto.getDefaultParams(self)
    }
}

// swift-migration: original location ARTCrypto.h, line 38 and ARTCrypto.m, line 39
public class ARTCipherParams: NSObject, ARTCipherParamsCompatible {
    
    // swift-migration: original location ARTCrypto.h, line 43
    public let algorithm: String
    // swift-migration: original location ARTCrypto.h, line 48
    public let key: Data
    // swift-migration: original location ARTCrypto.h, line 53
    public let keyLength: UInt
    // swift-migration: original location ARTCrypto+Private.h, line 9
    internal let iv: Data?
    
    // swift-migration: original location ARTCrypto.h, line 58
    public var mode: String {
        return getMode()
    }
    
    // swift-migration: original location ARTCrypto.h, line 64 and ARTCrypto.m, line 41
    public init(algorithm: String, key: ARTCipherKeyCompatible) {
        let keyData = key.toData()
        self.algorithm = algorithm
        self.key = keyData
        self.keyLength = UInt(keyData.count * 8)
        self.iv = nil
        super.init()
        
        var ccAlgorithm: CCAlgorithm = 0
        do {
            try self.ccAlgorithm(&ccAlgorithm)
        } catch {
            fatalError(error.localizedDescription)
        }
    }
    
    // swift-migration: original location ARTCrypto+Private.h, line 10 and ARTCrypto.m, line 46
    internal init(algorithm: String, key: ARTCipherKeyCompatible, iv: Data?) {
        let keyData = key.toData()
        self.algorithm = algorithm
        self.key = keyData
        self.keyLength = UInt(keyData.count * 8)
        self.iv = iv
        super.init()
        
        var ccAlgorithm: CCAlgorithm = 0
        do {
            try self.ccAlgorithm(&ccAlgorithm)
        } catch {
            fatalError(error.localizedDescription)
        }
    }
    
    // swift-migration: original location ARTCrypto.m, line 63
    private func getMode() -> String {
        return "CBC"
    }
    
    // swift-migration: original location ARTCrypto.m, line 8 and ARTCrypto.m, line 67
    internal func ccAlgorithm(_ algorithm: inout CCAlgorithm) throws {
        var errorMsg: String?
        
        if self.algorithm.caseInsensitiveCompare("AES") == .orderedSame {
            if let iv = self.iv, iv.count != cbcBlockLength {
                errorMsg = "iv length expected to be \(cbcBlockLength), got \(iv.count) instead"
            } else if self.keyLength != 128 && self.keyLength != 256 {
                errorMsg = "invalid key length for AES algorithm: \(self.keyLength)"
            } else {
                algorithm = CCAlgorithm(kCCAlgorithmAES)
            }
        } else if self.algorithm.caseInsensitiveCompare("DES") == .orderedSame {
            algorithm = CCAlgorithm(kCCAlgorithmDES)
        } else if self.algorithm.caseInsensitiveCompare("3DES") == .orderedSame {
            algorithm = CCAlgorithm(kCCAlgorithm3DES)
        } else if self.algorithm.caseInsensitiveCompare("CAST") == .orderedSame {
            algorithm = CCAlgorithm(kCCAlgorithmCAST)
        } else if self.algorithm.caseInsensitiveCompare("RC4") == .orderedSame {
            algorithm = CCAlgorithm(kCCAlgorithmRC4)
        } else if self.algorithm.caseInsensitiveCompare("RC2") == .orderedSame {
            algorithm = CCAlgorithm(kCCAlgorithmRC2)
        } else {
            errorMsg = "unknown algorithm: \(self.algorithm)"
        }
        
        if let errorMsg = errorMsg {
            throw NSError(domain: ARTAblyErrorDomain, code: 0, userInfo: [NSLocalizedFailureReasonErrorKey: errorMsg])
        }
    }
    
    // swift-migration: original location ARTCrypto.h, line 67 and ARTCrypto.m, line 99
    public func toCipherParams() -> ARTCipherParams {
        return self
    }
}

// swift-migration: original location ARTCrypto+Private.h, line 14
internal protocol ARTChannelCipher {
    func encrypt(_ plaintext: Data, output: inout Data?) -> ARTStatus
    func decrypt(_ ciphertext: Data, output: inout Data?) -> ARTStatus
    var cipherName: String? { get }
    var keyLength: Int { get }
}

// swift-migration: original location ARTCrypto+Private.h, line 23 and ARTCrypto.m, line 113
internal class ARTCbcCipher: NSObject, ARTChannelCipher {
    
    // swift-migration: original location ARTCrypto+Private.h, line 30
    internal let keySpec: Data
    // swift-migration: original location ARTCrypto+Private.h, line 31
    internal var iv: Data?
    // swift-migration: original location ARTCrypto+Private.h, line 32
    internal let blockLength: UInt
    // swift-migration: original location ARTCrypto+Private.h, line 29
    internal let logger: ARTInternalLog
    // swift-migration: original location ARTCrypto.m, line 16
    private let algorithm: CCAlgorithm
    
    // swift-migration: original location ARTCrypto+Private.h, line 25 and ARTCrypto.m, line 115
    internal init(cipherParams: ARTCipherParams, logger: ARTInternalLog) throws {
        self.keySpec = cipherParams.key
        self.iv = cipherParams.iv
        self.blockLength = UInt(cbcBlockLength)
        self.logger = logger
        
        var ccAlgorithm: CCAlgorithm = 0
        try cipherParams.ccAlgorithm(&ccAlgorithm)
        self.algorithm = ccAlgorithm
        
        super.init()
    }
    
    // swift-migration: original location ARTCrypto+Private.h, line 19 and ARTCrypto.m, line 130
    internal var keyLength: Int {
        return keySpec.count * 8
    }
    
    // swift-migration: original location ARTCrypto+Private.h, line 26 and ARTCrypto.m, line 134
    class func cbcCipher(params: ARTCipherParams, logger: ARTInternalLog) throws -> ARTCbcCipher {
        return try ARTCbcCipher(cipherParams: params, logger: logger)
    }
    
    // swift-migration: original location ARTCrypto+Private.h, line 16 and ARTCrypto.m, line 138
    internal func encrypt(_ plaintext: Data, output: inout Data?) -> ARTStatus {
        guard let iv = self.iv ?? ARTCrypto.generateSecureRandomData(Int(self.blockLength)) else {
            ARTLogError(self.logger, "ARTCrypto error encrypting")
            return ARTStatus(state: .error)
        }
        
        // The maximum cipher text is plaintext length + block length. We are also prepending this with the IV so need 2 block lengths in addition to the plaintext length.
        let outputBufLen = plaintext.count + Int(self.blockLength) * 2
        let buf = UnsafeMutableRawPointer.allocate(byteCount: outputBufLen, alignment: 1)
        
        // Copy the iv first
        iv.withUnsafeBytes { ivBytes in
            buf.copyMemory(from: ivBytes.baseAddress!, byteCount: Int(self.blockLength))
        }
        
        let ciphertextBuf = buf.advanced(by: Int(self.blockLength))
        let ciphertextBufLen = outputBufLen - Int(self.blockLength)
        
        var bytesWritten: size_t = 0
        let status = keySpec.withUnsafeBytes { keyBytes in
            iv.withUnsafeBytes { ivBytes in
                plaintext.withUnsafeBytes { plaintextBytes in
                    CCCrypt(
                        CCOperation(kCCEncrypt),
                        algorithm,
                        CCOptions(kCCOptionPKCS7Padding),
                        keyBytes.baseAddress,
                        keySpec.count,
                        ivBytes.baseAddress,
                        plaintextBytes.baseAddress,
                        plaintext.count,
                        ciphertextBuf,
                        ciphertextBufLen,
                        &bytesWritten
                    )
                }
            }
        }
        
        if status != kCCSuccess {
            ARTLogError(self.logger, "ARTCrypto error encrypting. Status is \(status)")
            buf.deallocate()
            return ARTStatus(state: .error)
        }
        
        let ciphertext = Data(bytesNoCopy: buf, count: bytesWritten + Int(self.blockLength), deallocator: .free)
        if ciphertext.isEmpty {
            ARTLogError(self.logger, "ARTCrypto error encrypting. cipher text is nil")
            buf.deallocate()
            return ARTStatus(state: .error)
        }
        
        output = ciphertext
        return ARTStatus(state: .ok)
    }
    
    // swift-migration: original location ARTCrypto+Private.h, line 17 and ARTCrypto.m, line 185
    internal func decrypt(_ ciphertext: Data, output: inout Data?) -> ARTStatus {
        // The first *blockLength* bytes are the iv
        if ciphertext.count < Int(self.blockLength) {
            return ARTStatus(state: .invalidArgs)
        }
        
        let ivData = ciphertext.subdata(in: 0..<Int(self.blockLength))
        let actualCiphertext = ciphertext.subdata(in: Int(self.blockLength)..<ciphertext.count)
        
        let options: CCOptions = 0
        
        // The output will never be more than the input + block length
        let outputLength = actualCiphertext.count + Int(self.blockLength)
        let buf = UnsafeMutableRawPointer.allocate(byteCount: outputLength, alignment: 1)
        var bytesWritten: size_t = 0
        
        // Decrypt without padding because CCCrypt does not return an error code
        // if the decrypted value is not padded correctly
        let status = keySpec.withUnsafeBytes { keyBytes in
            ivData.withUnsafeBytes { ivBytes in
                actualCiphertext.withUnsafeBytes { ciphertextBytes in
                    CCCrypt(
                        CCOperation(kCCDecrypt),
                        algorithm,
                        options,
                        keyBytes.baseAddress,
                        keySpec.count,
                        ivBytes.baseAddress,
                        ciphertextBytes.baseAddress,
                        actualCiphertext.count,
                        buf,
                        outputLength,
                        &bytesWritten
                    )
                }
            }
        }
        
        if status != kCCSuccess {
            ARTLogError(self.logger, "ARTCrypto error decrypting. Status is \(status)")
            buf.deallocate()
            return ARTStatus(state: .error)
        }
        
        // Check that the decrypted value is padded correctly and determine the unpadded length
        let cbuf = buf.bindMemory(to: Int8.self, capacity: bytesWritten)
        let paddingLength = Int(cbuf[bytesWritten - 1])
        
        if paddingLength == 0 || paddingLength > bytesWritten {
            buf.deallocate()
            return ARTStatus(state: .cryptoBadPadding)
        }
        
        for i in stride(from: bytesWritten - 1, to: bytesWritten - paddingLength, by: -1) {
            if paddingLength != Int(cbuf[i - 1]) {
                buf.deallocate()
                return ARTStatus(state: .cryptoBadPadding)
            }
        }
        
        let unpaddedLength = bytesWritten - paddingLength
        let plaintext = Data(bytesNoCopy: buf, count: unpaddedLength, deallocator: .free)
        if plaintext.isEmpty {
            ARTLogError(self.logger, "ARTCrypto error decrypting. plain text is nil")
            buf.deallocate()
        }
        
        output = plaintext
        return ARTStatus(state: .ok)
    }
    
    // swift-migration: original location ARTCrypto+Private.h, line 18 and ARTCrypto.m, line 250
    internal var cipherName: String? {
        let algo: String
        switch algorithm {
        case CCAlgorithm(kCCAlgorithmAES):
            algo = "aes"
        case CCAlgorithm(kCCAlgorithmDES):
            algo = "des"
        case CCAlgorithm(kCCAlgorithm3DES):
            algo = "3des"
        case CCAlgorithm(kCCAlgorithmCAST):
            algo = "cast"
        case CCAlgorithm(kCCAlgorithmRC4):
            algo = "rc4"
        case CCAlgorithm(kCCAlgorithmRC2):
            algo = "rc2"
        default:
            fatalError("Invalid algorithm")
        }
        return "\(algo)-cbc"
    }
}

// swift-migration: original location ARTCrypto.h, line 74 and ARTCrypto.m, line 280
public class ARTCrypto: NSObject {
    
    // swift-migration: original location ARTCrypto+Private.h, line 38 and ARTCrypto.m, line 282
    public class func defaultAlgorithm() -> String {
        return "AES"
    }
    
    // swift-migration: original location ARTCrypto+Private.h, line 39 and ARTCrypto.m, line 286
    internal class func defaultKeyLength() -> Int {
        return 256
    }
    
    // swift-migration: original location ARTCrypto+Private.h, line 40 and ARTCrypto.m, line 290
    internal class func defaultBlockLength() -> Int {
        return 128
    }
    
    // swift-migration: original location ARTCrypto+Private.h, line 42 and ARTCrypto.m, line 294
    internal class func generateSecureRandomData(_ length: Int) -> Data? {
        var data = Data(count: length)
        let result = data.withUnsafeMutableBytes { bytes in
            SecRandomCopyBytes(kSecRandomDefault, length, bytes.baseAddress!)
        }
        return result == errSecSuccess ? data : nil
    }
    
    // swift-migration: original location ARTCrypto+Private.h, line 43 and ARTCrypto.m, line 312
    internal class func generateHashSHA256(_ data: Data) -> Data {
        var digest = [UInt8](repeating: 0, count: Int(CC_SHA256_DIGEST_LENGTH))
        _ = data.withUnsafeBytes { bytes in
            CC_SHA256(bytes.baseAddress, CC_LONG(data.count), &digest)
        }
        return Data(digest)
    }
    
    // swift-migration: original location ARTCrypto.h, line 82 and ARTCrypto.m, line 320
    public class func getDefaultParams(_ cipherParams: [String: Any]) -> ARTCipherParams {
        let algorithm = cipherParams["algorithm"] as? String ?? ARTCrypto.defaultAlgorithm()
        guard let key = cipherParams["key"] as? String else {
            fatalError("missing key parameter")
        }
        return ARTCipherParams(algorithm: algorithm, key: key)
    }
    
    // swift-migration: original location ARTCrypto.h, line 99 and ARTCrypto.m, line 332
    public class func generateRandomKey() -> Data? {
        return generateRandomKey(UInt(defaultKeyLength()))
    }
    
    // swift-migration: original location ARTCrypto.h, line 91 and ARTCrypto.m, line 336
    public class func generateRandomKey(_ length: UInt) -> Data? {
        return generateSecureRandomData(Int(length / 8))
    }
    
    // swift-migration: original location ARTCrypto+Private.h, line 45 and ARTCrypto.m, line 340
    internal class func cipher(params: ARTCipherParams, logger: ARTInternalLog) throws -> ARTChannelCipher {
        return try ARTCbcCipher.cbcCipher(params: params, logger: logger)
    }
}